sources/rest_api/__init__.py

```diff
@@ -0,0 +1,768 @@
+"""Generic API Source"""
+
+from dataclasses import dataclass
+import copy
+from typing import Any, Dict, NamedTuple, Optional, TypedDict, Union
+
+import dlt
+import graphlib
+from dlt.common.schema.typing import (
+    TColumnNames,
+    # TSchemaContract,
+    TTableFormat,
+    TTableSchemaColumns,
+    TWriteDisposition,
+)
+from dlt.extract.incremental import Incremental
+from dlt.extract.source import DltResource
+from dlt.extract.typing import TTableHintTemplate
+
+from .auth import BearerTokenAuth, AuthBase
+from .client import RESTClient
+from .paginators import (
+    BasePaginator,
+    HeaderLinkPaginator,
+    JSONResponsePaginator,
+    UnspecifiedPaginator,
+    SinglePagePaginator,
+)
+from .utils import remove_key, deep_merge
+
+
+PAGINATOR_MAP = {
+    "json_links": JSONResponsePaginator,
+    "header_links": HeaderLinkPaginator,
+    "auto": UnspecifiedPaginator,
+    "single_page": SinglePagePaginator,
+}
+
+
+PaginatorType = Union[str, BasePaginator]
+
+
+class AuthConfig(TypedDict, total=False):
+    token: str
+
+
+class ClientConfig(TypedDict, total=False):
+    base_url: str
+    auth: Optional[AuthConfig]
+    default_paginator: Optional[PaginatorType]
+
+
+class ResourceConfig(TypedDict, total=False):
+    primary_key: str
+    write_disposition: str
+
+
+class IncrementalConfig(TypedDict, total=False):
+    cursor_path: str
+    initial_value: str
+    param: str
+
+
+class EndpointConfig(TypedDict):
+    params: Dict[str, Any]
+    resource: ResourceConfig
+    incremental: Optional[IncrementalConfig]
+    method: str
+    paginator: Optional[PaginatorType]
+
+
+class RESTAPIConfigLegacy(TypedDict):
+    client: ClientConfig
+    endpoints: Dict[str, EndpointConfig]
+
+
+class ResolveConfig(NamedTuple):
+    resource_name: str
+    field_path: str
+
+
+class ResolvedParam(NamedTuple):
+    param_name: str
+    resolve_config: ResolveConfig
+
+
+class Endpoint(TypedDict, total=False):
+    path: str
+    method: str
+    params: Optional[Dict[str, Any]]
+    json: Optional[Dict[str, Any]]
+    paginator: Optional[PaginatorType]
```

Comment:
Ideally we should also allow instance of paginator here (and maybe later a dict with paginator config)